### ISSUE TITLE
Alternator test runs extremely slowly on Jenkins machines

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#6750]
Recently some next promotion runs failed on the Alternator tests (test/alternator), e.g., https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/2172/consoleFull

The failure is always in the setup of the `test_lsi_4` fixture (used from two tests), when creating the test table - Alternator claims the table already exists:
```
botocore.errorfactory.ResourceInUseException: An error occurred (ResourceInUseException) when calling the CreateTable operation: Table alternator_Test_1593654675766 already exists
```
The table-name creation function `test_table_name()` names the table according to the current time in millisecond resolution, but has some logic on avoiding picking the same millsecond a second time. Could we have a bug in this logic?

CC @psarna

**Comment by avikivity** [comment]
Maybe boto3 is getting a false failure on table creation, and retrying. The retry fails since the table already exists.

**Comment by nyh** [comment]
The short story:
The test is running absurdly slowly, over 100 times slower than my laptop, which results in boto3 reaching a one minute (!) timeout and retrying the table creation. There are a lot of things we could do, but I think the main thing is to avoid this absurd slowness. @fruch - suggestions?

The full details:

Here are the entries in the log creating the problematic table alternator_Test_1593654750611:
```
WARN  2020-07-02 01:52:30,613 [shard 0] alternator-executor - Creating keyspace 'alternator_alternator_Test_1593654750611' for Alternator with unsafe RF=1 because cluster only has 1 nodes.
WARN  2020-07-02 01:53:30,856 [shard 1] alternator-executor - Creating keyspace 'alternator_alternator_Test_1593654750611' for Alternator with unsafe RF=1 because cluster only has 1 nodes.
```

Avi's guess was right on the money. It's not a coincidence that the second table creation is almost exactly one minute after the first one. I have seen in the past that boto3 has a one-minute timeout for hung requests: If it sends a request and doesn't get any reply in one minute, it closes the connection and tries again.

I believe this is exactly what happened in this case. Looking at the logs between 01:52:30 and 1:53:30, I see mainly a lot of compactions, view buildings, etc, all part of the origin CreateTable request. As late as 01:53:42,200 - i.e., after the second attempt to create - I still see us trying to create materialized views on this table, which are still part of the original CreateTable attempt!

So I think this bug is a combination of:
1. An *extremely slow* run. Yes, test_table_lsi_4 creation is a lot of work (a base table and 4 materialized views), but shouldn't take a full minute unless the host is super-overcommitted or a Commodore 64. On my machine, this test takes just 0.2 seconds!
2. Issue #5052 - CreateTable should not wait for the table and all of its materialized views - it should return relatively fast and boto3 would not have retried. We need to fix this issue, but I was hoping that we can postpone fixing it indefinitely (because who cares about waiting a fraction of a second?), but if we have servers hundreds of times slower than usual, we can't....
3. test_table_lsi_4 is the biggest table-creation in this test, with 4 materialized views. This is why we're always seeing failure in this specific case every time - and not in other random tests, which are shorter (and finish in less than a full minute).

@avikivity @fruch the test machine slowness is a main factor in this failure. Although I should indeed fix #5052, I think the slowness is much worse and we must take care of it. One obvious sign of the slowness is this log line:

```
Booting Scylla.........................................Done (44 seconds)
```

On my laptop, it's just 1 (one) second!! It's actually even less than one second - it's rounded up to full seconds :-(

**Comment by nyh** [comment]
@avikivity the Alternator test (test/alternator/run) runs Scylla with "--overprovisioned". Could it be that this option is allowing other concurrent tests (There are, right?) to take most of the CPU, resulting in the Alternator test running super slowly? @fruch is there anything in the test log I can use to figure out how many tests were running in parallel to this test?
Another possibility we have seen in the past is that sometimes we get VMs running qemu **without** KVM, and are a hundred times slower. Is this even a possibility in our Jenkins setup, or are we not using VMs?

**Comment by nyh** [comment]
Question for @fruch: @psarna also suggest that maybe: 1. we're running many other tests in parallel (are we?), and 2. maybe some other tests got a higher "nice" priority than the Alternator test, so the Alternator test is running much less than it should?

And a question for @avikivity: Is it possible that with "--overprovisioned", every time that Scylla needs to wait a microsecond for disk activity, it now waits a whole timeslice (for another test to run), and because the boot and tests have zero concurrency, this is slowing them down considerably? Do we do anything in dtests or unit tests to avoid this problem?

**Comment by fruch** [comment]
seem like this is running on spot fleet (it should have 30 cpus on those)
https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2172

and maybe we are running 3 builds at the same time, and you happen to run
while scylla is building

@Yaron Kaikov <yaronkaikov@scylladb.com> Is this setup now building
parallelly on the same machine ? or running some test parallely ? like
dtest and unittest ?

**Comment by yaronkaikov** [comment]
@fruch No, it's running on spot but same as local machines. Parallel is not in yet

**Comment by nyh** [comment]
@yarongilor it's not parallel *at all*? There is absolutely nothing running in parallel with the Alternator test? Not even other tests?

@fruch @yaronkaikov can one of you try to run the Alternator test manually on this machine to try to get a feeling what could possibly make it hundreds of times (not just 2 times) slower than on my laptop?

If I need to fix this test to be fine on 200-times-slower nodes, I will, but really, I don't think we should have 200-times-slower nodes at all.

**Comment by avikivity** [comment]
Other tests are running in parallel, but not more than the cpu count. So there should be ample cpu bandwidth.

@yaronkaikov what is the instance configuration? what is the disk used?

**Comment by avikivity** [comment]
I see it is a gp2 disk with 900 iops.

**Comment by yaronkaikov** [comment]
@avikivity for this build it used r3.8xlarge which have 32 vCPUs and 244 Gb RAM , uses SSD

**Comment by yaronkaikov** [comment]
@avikivity it is configured to gp2, i think it is using the configuration for the AMI i am using. I can modify it to io1 which is also SSD that provides upto 64,000 IOPS

**Comment by nyh** [comment]
> Other tests are running in parallel, but not more than the cpu count. So there should be ample cpu bandwidth.

If there are 32 vCPUs, it means 32 tests are running in parallel?
Could it be that some of these tests use many CPUs, not just one? The Alternator test uses two, which isn't a big problem, but could we have tests that use **all** CPUs?

**Comment by avikivity** [comment]
@yaronkaikov better to use the local storage instead.

**Comment by avikivity** [comment]
@nyh no, our tests run with --smp 1 or --smp 2

Note that those tests passed in debug and release mode. It's not cpu overcommit.

**Comment by avikivity** [comment]
Failed again:

```
17:35:05  [242/426] boost/range_tombstone_list_test                      dev    [ PASS ]
17:35:10  [243/426] boost/querier_cache_test                             dev    [ PASS ]
17:35:10  [244/426] boost/reusable_buffer_test                           dev    [ PASS ]
17:35:12  [245/426] boost/reusable_buffer_test                           dev    [ PASS ]
17:35:12  [246/426] boost/reusable_buffer_test                           dev    [ PASS ]
17:35:20  [247/426] alternator/run                                       dev    [ FAIL ]
17:35:23  [248/426] boost/querier_cache_test                             dev    [ PASS ]
17:35:28  [249/426] alternator/run                                       dev    [ FAIL ]
17:35:45  [250/426] boost/cdc_test                                       dev    [ PASS ]
17:35:46  [251/426] boost/querier_cache_test                             dev    [ PASS ]
17:35:48  [252/426] boost/row_cache_test                                 dev    [ PASS ]
17:35:50  [253/426] alternator/run                                       dev    [ FAIL ]
17:35:51  [254/426] boost/row_cache_test                                 dev    [ PASS ]
17:35:52  [255/426] boost/schema_registry_test                           dev    [ PASS ]
```

Very suspicious - debug mode and release mode pass. So something's not working right.

**Comment by nyh** [comment]
@avikivity why do I see here each test runs 3 times, and all in parallel?
Can we figure out how long the other tests take? Maybe they are all 100 times slower than expected?
The alternator test has a particular problem with the 1 minute timeout that breaks it, but it's possible that all tests suffer from the extreme slowdown.

**Comment by yaronkaikov** [comment]
@avikivity modified the disk with our next spot instances and now tests are passing https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2176/ . although dtest is still failing

**Comment by nyh** [comment]
Yaron, how much time does the Alternator test take now? On my laptop it takes 25 seconds...

**Comment by yaronkaikov** [comment]
Not sure... https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2176/testReport/(root)/alternator/  - is this the right place ?

**Comment by nyh** [comment]
> Not sure... https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2176/testReport/(root)/alternator/  - is this the right place ?
> 

Unrelated to this issue, but why does that page only show 27 tests and not the more than 400 we actually have in alternator tests? And least times like "0ms" look encouraging.

**Comment by nyh** [comment]
> @avikivity modified the disk with our next spot instances and now tests are passing https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2176/ . although dtest is still failing

I see that the "Test" (unit tests - test.py) time in 2176 is down to 2 hours 39 minutes, from 3 hours 51 minutes in 2172. That's a good improvement, but I can't figure out how if the ridiculous slowness of the Alternator test was really solved, or simply the slow table create went down from a little over one minute to a little less than one minute, so the bug disappeared while the ridiculous slowness remains.

Unfortunately, test.py prints lines like:

```
01:40:58  [271/426] boost/row_cache_test                                 dev    [ PASS ]
01:41:03  [272/426] alternator/run                                       dev    [ PASS ]
01:41:23  [273/426] boost/multishard_mutation_query_test                 dev    [ PASS ]
01:41:36  [274/426] boost/cql_query_like_test                            dev    [ PASS ]
```
which aren't helpful at all to understand how much time each take took to run.
@kostja what is your opinion about after the '[ PASS ]" string also the amount of time the test took, from start to finish? Again, alternator/run takes on my own laptop just 27 seconds - if I see there "27 seconds" I'll be very happy, if I see "100 seconds" I'll still be happy (CPU overcommit can explain it), but if I see "27 minutes" I'll know something really really bad happened.

**Comment by kostja** [comment]
@nyh we have discussed this earlier. Alternator should not use a shell script to wrap & hide pytest tests. Each pytest should be visible to test.py as an own test. I think the suggestion to add time to the tabular output while is harmless but is fixing a wrong problem.

**Comment by kostja** [comment]
I hope I will have time to add direct pytest support to test.py in scope of work on schema changes tests, then alternator will be able to switch.

**Comment by nyh** [comment]
> @nyh we have discussed this earlier. Alternator should not use a shell script to wrap & hide pytest tests. Each pytest should be visible to test.py as an own test. I think the suggestion to add time to the tabular output while is harmless but is fixing a wrong problem.

You are mixing up two unrelated issues....
One is that even if Alternator's 500 (!) different tests, or 26 test files, would have each been considered by test.py a "separate test", I would still like to know that each of those took 10 seconds instead of 10 milliseconds - and the current log format doesn't let me know that... It only says when each test was started, but because of the parallelism I have no idea when each test finished.
The second thing is that you can't split Alternator's 500 different tests, or even 26 test files, into separate runs, because it will make these runs much much slower because we could no longer reuse the same "fixtures" (temporary tables, etc., or even the running Scylla). As I repeatedly explained, right now all 500 Alternator tests finish in as little as 25 seconds (on one CPU on my laptop). Making them take 25 minutes - like dtests - is the last thing I want to see happen.

**Comment by nyh** [comment]
> @avikivity why do I see here each test runs 3 times,

Wow, I just realized we use "--repeat 3" when we run tests. I'm stumped. But it seems deliberate.

**Comment by kostja** [comment]
> > @nyh we have discussed this earlier. Alternator should not use a shell script to wrap & hide pytest tests. Each pytest should be visible to test.py as an own test. I think the suggestion to add time to the tabular output while is harmless but is fixing a wrong problem.
> 
> You are mixing up two unrelated issues....
> 
> One is that even if Alternator's 500 (!) different tests, or 26 test files, would have each been considered by test.py a "separate test", I would still like to know that each of those took 10 seconds instead of 10 milliseconds - and the current log format doesn't let me know that... It only says when each test was started, but because of the parallelism I have no idea when each test finished.

The time it takes for a test to run is present in test log output that test.py grabs and saves, as well as in xunit xml report.

> The second thing is that you can't split Alternator's 500 different tests, or even 26 test files, into separate runs, because it will make these runs much much slower because we could no longer reuse the same "fixtures" (temporary tables, etc., or even the running Scylla). As I repeatedly explained, right now all 500 Alternator tests finish in as little as 25 seconds (on one CPU on my laptop). Making them take 25 minutes - like dtests - is the last thing I want to see happen.

pytest tests shouldn't lead to scylla restart.

**Comment by avikivity** [comment]
> > Not sure... https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2176/testReport/(root)/alternator/  - is this the right place ?
> 
> Unrelated to this issue, but why does that page only show 27 tests and not the more than 400 we actually have in alternator tests? And least times like "0ms" look encouraging.

"0ms" is an indication that the test didn't run, more than it was fast.

**Comment by nyh** [comment]
> > > Not sure... https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2176/testReport/(root)/alternator/  - is this the right place ?
> > 
> > 
> > Unrelated to this issue, but why does that page only show 27 tests and not the more than 400 we actually have in alternator tests? And least times like "0ms" look encouraging.
> 
> "0ms" is an indication that the test didn't run, more than it was fast.

Avi, I don't understand why are you saying this. These tests really can take less than a millisecond each, especially in release or dev mode... Each test runs 9 times (3 build modes, 3 repetitions each). It makes a lot of sense that the debug mode runs took 2 or 1 milliseconds, the release mode took 0 (less than 1) millisecond. 

None of this has anything to do with the real problem of why do we see only three sub-tests in this list, not all of them. Maybe I did something wrong with the pytest's "--junit-xml" output flag? But I don't see where I got it wrong, if at all.

**Comment by nyh** [comment]
> > > > Not sure... https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2176/testReport/(root)/alternator/  - is this the right place ?
> > > 
> > > 
> > > Unrelated to this issue, but why does that page only show 27 tests and not the more than 400 we actually have in alternator tests? And least times like "0ms" look encouraging.

Turns out the three tests (each repeated 9 times - 3 build modes times 3 repetitions) are NOT from `test/alternator`, but three tests from `test/boost/alternator_base64_test.cc`. So we don't see the Alternator tests at all, despite passing "--junit-xml" to test.py. Now I need to figure out why. I'll open a separate issue on that.

**Comment by nyh** [comment]
https://github.com/scylladb/scylla/issues/6767

**Comment by nyh** [comment]
I'm closing this issue because it has turned into a bunch of ideas and the original complained (about Alternator test failing) doesn't happen any more, and I opened other more specific issues - #5052, #6759, #6767.

Still, @yaronkaikov @fruch I urge you to please look into **how much** tests are slowed down by the test setup. Slowing tests 2 or even 10-fold compared to developers' laptops is perfectly acceptable. But slowing them 100-fold or more (if it's happening...) is a problem because it will surely lead to spurious failures like in this issue.

**Comment by avikivity** [comment]
> > > > Not sure... https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2176/testReport/(root)/alternator/  - is this the right place ?
> > > 
> > > 
> > > Unrelated to this issue, but why does that page only show 27 tests and not the more than 400 we actually have in alternator tests? And least times like "0ms" look encouraging.
> > 
> > 
> > "0ms" is an indication that the test didn't run, more than it was fast.
> 
> Avi, I don't understand why are you saying this. These tests really can take less than a millisecond each, especially in release or dev mode... Each test runs 9 times (3 build modes, 3 repetitions each). It makes a lot of sense that the debug mode runs took 2 or 1 milliseconds, the release mode took 0 (less than 1) millisecond.

1ms is pretty fast for a test involving networking and possibly storage, even if local.

**Comment by nyh** [comment]
Seen again in a different Jenkins run, https://jenkins.scylladb.com/view/master/job/scylla-master/job/releng-testing/job/next/100/console
Again, lsi_4 fixture failed to create table in one minute because of super-slow run, saw this in the log
```
Booting Scylla..........................................Done (43 seconds)"
```
(compared to 1 second on my laptop, which actually means less than 1 second).

Again, this is happening **only** for dev-mode run, I have no explanation why (debug mode should have been even slower).

I asked @yaronkaikov to check in a manual run (`test.py --mode=dev alternator/run`) on the same VM whether it is also so slow when running alone, or something in the way that 30 different tests are run together causes this slowness. As a comparison, on my laptop, this test finishes in less than 30 seconds.

**Comment by yaronkaikov** [comment]
@nyh So i have run this test manually using the same command (with dbuild) and i see the slowness on the AWS machines **although they are passing and not failing**
[jenkins@ip-10-0-17-112 scylla]$ time ./tools/toolchain/dbuild --privileged -e UBSAN_OPTIONS=print_stacktrace=1 -- ./test.py --mode=dev --verbose --timeout 7200 --repeat 3 alternator/run
================================================================================
[N/TOTAL] TEST                                                 MODE   RESULT  
------------------------------------------------------------------------------
[1/3]     alternator/run                                       dev    [ PASS ]
[2/3]     alternator/run                                       dev    [ PASS ]
[3/3]     alternator/run                                       dev    [ PASS ]
------------------------------------------------------------------------------

real	1m25.953s
user	0m0.137s
sys	0m0.105s

**Comment by nyh** [comment]
That's not the slowness, that's fastness 😂 running the test 3 times took 1 and a half minutes. That's 30 seconds for each run!
Compare this to the run on Jenkins which I can't even guess how long everything took, but I know that just booting Scylla took 46 seconds and then juzt one test, taking 0.2 seconds on my laptop, timed out after a minute.

**Comment by nyh** [comment]
I'm reopening, since the problem returned as in the link above and also in yet another run:
https://jenkins.scylladb.com/view/master/job/scylla-master/job/releng-testing/job/next/109/consoleFull

The problem is the same every time - Alternator test runs extremely slowly, to the point that just booting Scylla - something which normally takes less than a second, takes 46 seconds, and later the lsi_4 table creation (a table plus four views) - which normally takes 0.2 seconds - takes more than a minute and times out, failing the test. The most unexpected - but consistent - observation is that it is always the dev mode that fails, no other mode - even though the debug mode is usually significantly slower.

We are stumped what makes Alternator run Scylla so extremely slowly. There are endless possibilities, but I have no idea how to check them without having access or even knowledge on how does the test machine run. Here are some guesses:

1. Does the test machine do "only" 32 tests concurrently (for 32 cores), or does it do a bunch other stuff concurrently? In particular, since it is *always* the dev mode which fails, not the other modes, is it possible that we do other things in parallel to the dev mode run but not other runs? Perhaps we do other builds in parallel?
2. Are we sure the test machine does not do swapping? This can easily make things 100 times slower.
3. How many IOPS can the test machine do? If it has just low IOPS, it can explain why running 32 tests in parallel is very slow.
4. Alternator test runs Scylla in "overcommit" mode, which can easily relinquish its cpu quota to other threads. Could it be that this is why it is loosing CPU to other concurrent tests? Do we have anything running on this machine with increased "niceness" that can cause it to monopolize the CPU?

As @yaronkaikov noted above, when the test is run alone - not together with the entire Jenkins run - it finishes very fast - 30 seconds for the entire thing per run, which includes the Scylla boot and 500 tests. Compare this to 46 seconds for boot and more than a minute for a single one of those 500 tests in the slow run that we don't understand....

**Comment by yaronkaikov** [comment]
I run those tests (part of next-debug process) 3 times . we are using here AWS spot instance with local SSDs. 
On first run i set the repeat to 1, everything passed:
https://jenkins.scylladb.com/view/master/job/scylla-master/job/releng-testing/job/next/111/
Second run with test repeat = 2, all passed: 
https://jenkins.scylladb.com/view/master/job/scylla-master/job/releng-testing/job/next/112/
Third run with test repeat = 3, alternator/run failed again on dev mode only (there was another test that filed during this run):
https://jenkins.scylladb.com/view/master/job/scylla-master/job/releng-testing/job/next/113/

Booting Scylla was 41 sec 
05:22:18  Scylla log:
05:22:18  test/alternator/run: line 40:   136 Killed                  "$SCYLLA_LINK" --options-file "$source_path/conf/scylla.yaml" --alternator-address $SCYLLA_IP $alternator_port_option --alternator-enforce-authorization=1 --alternator-write-isolation=always_use_lwt --developer-mode=1 --ring-delay-ms 0 --collectd 0 --smp 2 -m 1G --overprovisioned --unsafe-bypass-fsync 1 --api-address $SCYLLA_IP --rpc-address $SCYLLA_IP --listen-address $SCYLLA_IP --prometheus-address $SCYLLA_IP --seed-provider-parameters seeds=$SCYLLA_IP --workdir "$tmp_dir" --server-encryption-options keyfile="$tmp_dir/scylla.key" --server-encryption-options certificate="$tmp_dir/scylla.crt" --auto-snapshot 0 --skip-wait-for-gossip-to-settle 0 > "$tmp_dir/log" 2>&1  (wd: /jenkins/workspace/scylla-master/releng-testing/next/scylla)
05:22:18  Scylla version 666.development-0.20200712.ef2c8f563 with build-id c3f8f7994844e0c82eac89fefd4fa31320d7f984 starting ...
05:22:18  command used: "/tmp/alternator-test-9/test_scylla --options-file /jenkins/workspace/scylla-master/releng-testing/next/scylla/test/alternator/../../conf/scylla.yaml --alternator-address 127.1.0.9 --alternator-port=8000 --alternator-enforce-authorization=1 --alternator-write-isolation=always_use_lwt --developer-mode=1 --ring-delay-ms 0 --collectd 0 --smp 2 -m 1G --overprovisioned --unsafe-bypass-fsync 1 --api-address 127.1.0.9 --rpc-address 127.1.0.9 --listen-address 127.1.0.9 --prometheus-address 127.1.0.9 --seed-provider-parameters seeds=127.1.0.9 --workdir /tmp/alternator-test-9 --server-encryption-options keyfile=/tmp/alternator-test-9/scylla.key --server-encryption-options certificate=/tmp/alternator-test-9/scylla.crt --auto-snapshot 0 --skip-wait-for-gossip-to-settle 0"
05:22:18  parsed command line options: [options-file: /jenkins/workspace/scylla-master/releng-testing/next/scylla/test/alternator/../../conf/scylla.yaml, alternator-address: 127.1.0.9, alternator-port: 8000, alternator-enforce-authorization: 1, alternator-write-isolation: always_use_lwt, developer-mode: 1, ring-delay-ms: 0, collectd: 0, smp: 2, memory: 1G, overprovisioned, unsafe-bypass-fsync: 1, api-address: 127.1.0.9, rpc-address: 127.1.0.9, listen-address: 127.1.0.9, prometheus-address: 127.1.0.9, seed-provider-parameters: seeds=127.1.0.9, workdir: /tmp/alternator-test-9, server-encryption-options: keyfile=/tmp/alternator-test-9/scylla.key, server-encryption-options: certificate=/tmp/alternator-test-9/scylla.crt, auto-snapshot: 0, skip-wait-for-gossip-to-settle: 0]
05:22:18  INFO  2020-07-13 00:46:28,437 [shard 0] init - installing SIGHUP handler
05:22:18  INFO  2020-07-13 00:46:28,439 [shard 0] init - Scylla version 666.development-0.20200712.ef2c8f563 with build-id c3f8f7994844e0c82eac89fefd4fa31320d7f984 starting ...
05:22:18  WARN  2020-07-13 00:46:28,439 [shard 0] init - NOFILE rlimit too low (recommended setting 200000, minimum setting 10000; you may run out of file descriptors.
05:22:18  WARN  2020-07-13 00:46:28,439 [shard 0] init - Only 512 MiB per shard; this is below the recommended minimum of 1 GiB/shard; continuing since running in developer mode
05:22:18  INFO  2020-07-13 00:46:28,440 [shard 0] init - starting prometheus API server
05:22:18  INFO  2020-07-13 00:46:28,461 [shard 0] init - starting tokens manager
05:22:18  INFO  2020-07-13 00:46:28,461 [shard 0] init - starting migration manager notifier
05:22:18  INFO  2020-07-13 00:46:28,461 [shard 0] init - creating tracing
05:22:18  INFO  2020-07-13 00:46:28,461 [shard 0] init - creating snitch
05:22:18  INFO  2020-07-13 00:46:28,461 [shard 0] init - determining DNS name
05:22:18  INFO  2020-07-13 00:46:28,461 [shard 0] init - starting API server
05:22:18  INFO  2020-07-13 00:46:28,461 [shard 0] init - Scylla API server listening on 127.1.0.9:10000 ...
05:22:18  INFO  2020-07-13 00:46:28,462 [shard 0] init - initializing storage service
05:22:18  INFO  2020-07-13 00:46:28,462 [shard 0] init - starting per-shard database core
05:22:18  WARN  2020-07-13 00:46:28,464 [shard 0] init - I/O Scheduler is not properly configured! This is a non-supported setup, and performance is expected to be unpredictably bad.
05:22:18   Reason found: none of --max-io-requests, --io-properties and --io-properties-file are set.
05:22:18  To properly configure the I/O Scheduler, run the scylla_io_setup utility shipped with Scylla.

**Comment by nyh** [comment]
@yaronkaikov let me try to explain why you're seeing the test succeeding and failing in different runs.
The problematic test creates a table which normally (on my laptop) takes 0.2 seconds; But, if the table creation is slowed down to 60 seconds (!), then the DynamoDB Python driver gives up and the test fails.
In the failure I looked at a couple of weeks back, the table creation was slowed down to about 80 seconds (you can see when it started and ended in the failure log) - so because 80 > 60, the test failed. But it is quite possible that with some other random choice of parameters or alignment of the stars, the table creation would be slowed down to "only" 50 seconds, and since 50 < 60, the test would succeed. Still, 50 seconds is extremely slow - 250 times slower than the normal run, and if that happens we still need to understand why. We don't need to see a "failure" to see a problem - which is extreme slowness.

If we had test times in the log (#6759 - I sent a patch...) we could see the Alternator test run time even in successful runs. @xemul noted in his review of my patch that even before it, you can calculate the test's length yourself - look in `test.py.log` for the "starting test" line with `alternator/run`, note the test number, then look for the "Test ... succeeded" for this test's number, and subtract the times.

@yaronkaikov can you please try to claculate the `alternator/run` test time using this technique, on a successful run?
As you noted above, a good result would be 30 seconds. I could easily live with 2 or even 5 times more because of over-commitment of the test machine. But if it's 100 or 500 times slower, we have a serious problem that we need to solve.

**Comment by avikivity** [comment]
```sh
tmp_dir=/tmp/alternator-test-$$
mkdir $tmp_dir
```

This can place the temporary files on a slow disk (EBS) while $PWD is on a fast disk.

**Comment by nyh** [comment]
@avikivity, indeed Alternator tests currently uses /tmp for tests. In "normal" setups - such as development machines, /tmp is always the fastest possible disk (often, a ramdisk), are you saying that in our setup "/tmp" is an especially slow disk? Was this a deliberate choice? Do you know how slow it is (latency, IOPS)?

By the way, I asked @yaronkaikov to run the Alternator test alone (without fighting with any other test) on the same Jenkins setup, and seeing how fast or slowly it ran. This would have used the same slow /tmp (if this is true). He reported a 30 second run - which is very fast (same as on my laptop with a ramdisk for /tmp). But I'm not entriely convinced he actually ran it on exactly the same setup as Jenkins.

**Comment by avikivity** [comment]
Of course we did not deliberately choose to place /tmp on a slow disk. The Jenkins instance uses gp2 EBS, which provides 900 IOPS or so. There is also a much faster local disk, but it requires work to mount it.

**Comment by nyh** [comment]
Ok, I still don't understand why our Linux image doesn't use tmpfs (ram) for /tmp, like both Fedora and Ubunto have been doing since 2012, and I'm not convinced this is actually the problem (900 IOPS, if Alternator is the *only* test using /tmp, isn't entirely shabby), but let's work under this assumption that this is the problem.

I'll prepare a patch where test.py tells alternator/test/run where to put its temporary files.

**Comment by nyh** [comment]
@yaronkaikov @avikivity I sent a patch to the mailing list titled "alternator test: configurable temporary directory"
Please review it, and I'm hoping it fixes this issue (if it doesn't, we can always reopen it again...).

**Comment by nyh** [comment]
I just realized - by @espindola and @kostja making similar comments and TMPDIR patches - that actually many, if not most, unit tests use /tmp, not just Alternator's! So if 32 different tests are sharing a 900 IOPS disks concurrently, getting 30 IOPS each, it can easily explain the extreme slowness of all the tests.

Moreover, Ironically this can also explain why the dev build, not the debug build, is the slowest and causes most problems: When the 30 other competing tests run very quickly (i.e., not in debug mode), they use more I/O and make the Alternator run (which doesn't need much CPU) slower.

By the way, here are the logs, with new test times (finally my patch got in!), from a really fast test run https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2222/consoleFull - I'm not sure what is the difference in hardware that made this test fast - it's before my TMPDIR patch. Note that the "Test" part of the Jenkins pipeline was less than an hour, compared to more than 3 hours in other successful but slow runs. Anyway, here are the Alternator times:

```
18:22:33  [422/432] alternator/run                                     release  [ PASS ] 49.52s
18:22:33  [423/432] alternator/run                                     release  [ PASS ] 49.53s
18:22:33  [426/432] alternator/run                                     release  [ PASS ] 49.82s
18:39:09  [241/414] alternator/run                                      debug   [ PASS ] 959.31s
18:39:21  [242/414] alternator/run                                      debug   [ PASS ] 970.79s
18:39:39  [243/414] alternator/run                                      debug   [ PASS ] 989.46s
19:06:56  [414/432] alternator/run                                       dev    [ PASS ] 54.42s
19:06:56  [415/432] alternator/run                                       dev    [ PASS ] 54.83s
19:06:57  [417/432] alternator/run                                       dev    [ PASS ] 55.54s
```
These are all very reasonable numbers. Note how all of Alternator's 500 tests in dev mode finished in 54-55 seconds - while in the broken setup just a single test of those 500 timed out after 60 seconds...

**Comment by nyh** [comment]
I think we have a very good indication that /tmp was indeed the culprit. The run
https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2231/consoleFull
Had the commit "Export TMPDIR pointing at subdir of testlog/ " by @kostja. This didn't change the Alternator test yet (such a patch is already queued  as well - https://github.com/scylladb/scylla/commit/8e3be5e7d6046b357dcd0361df173dbe8b3aac37), that test still uses /tmp - but this patch moved all the other tests away from /tmp. And apparently not having to content with 30 other parallel tests all using /tmp concurrently did wonders for the Alternator test's performance - the entire test takes 70-500 seconds in this run, perfectly acceptable considering the test node is (by design) overcommitted. 
So I'm closing tis issue yet again.

**Comment by nyh** [comment]
It is not critical to backport this patch because 1. fixing the /tmp on Jenkins VMs will mean /tmp will be fine and a backport isn't needed. 2. If https://github.com/scylladb/scylla/commit/8e3be5e7d6046b357dcd0361df173dbe8b3aac37 is backported, it moves all the other tests except Alternator away from /tmp, which makes the Alternator tests - even if remaining on /tmp - much faster.
On the other hand, it doesn't hurt to backport this test-only patch either.

**Comment by nyh** [comment]
https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2235/consoleFull is the first run which came after both the TMPDIR patches, and ran on AWS. It shows reasonable times for the Alternator test - totalling (for the 500 subtests) about 50 seconds for the dev build, to 500 seconds for debug build.

**Comment by nyh** [comment]
Was already backported to 4.2.
Although it isn't strictly necessary (it's only a test improvement), to not risk test slowness I backported to 4.1 (32c0e4f11082257c610d32431a1bc3938ea1a2cf) and 4.0 (ada79df0829b6fa6d456d5c1e897a4cd7d148a8d)

**Comment by nyh** [comment]
We just had this problem of extreme slowness again - check out https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/2859/artifact/testlog/dev/run.1.log.
The tests no longer run on /tmp, they run on the current directory, but maybe the new machines have these slow disks we used to have for /tmp for the current directory?

**Comment by nyh** [comment]
We had the same failure again, in https://jenkins.scylladb.com/job/scylla-enterprise/job/next/1156/testReport/junit/test.alternator/test_update_expression/Test___test_update_expression_empty_attribute/
This time the failure was when trying to delete the test_table_lsi_4 table. Again, if boto3 times out the deletion call it retries it after a minute - and then this second attempt fails.
